### PR TITLE
Function1D's and fission Q-values

### PR DESCRIPTION
**Pull request by smharper**
This PR combines #695 and #690.  This also fixes a bug in #690 where the incident neutron energy wasn't subtracted from fission Q-values with Sher-Beck data.

Closes #695.  Closes #690

### CODE REVIEW

File: docs/source/io_formats/fission_energy.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+==================================
+Fission Energy Release File Format
+==================================
+
+This file is a compact HDF5 representation of the ENDF MT=1, MF=458 data (see
+ENDF-102_ for details).  It gives the information needed to compute the energy
+carried away from fission reactions by each reaction product (e.g. fragment
+nuclei, neutrons) which depends on the incident neutron energy.  OpenMC is
+distributed with one of these files under
+openmc/data/fission_Q_data_endfb71.h5.  More files of this format can be
```
Comment (paulromano) [inline docs/source/io_formats/fission_energy.rst]: Should be just data/fission_q_data_endfb71.h5, yeah?

**Comment by paulromano** [comment]
When I try to build the documentation on the branch, I get:
```
Exception occurred:
  File "/home/romano/openmc/openmc/data/function.py", line 358, in <module>
    class Polynomial(np.polynomial.Polynomial, Function1D):
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

File: scripts/openmc-ace-to-hdf5 [hunk 2]
```diff
@@ -111,6 +120,14 @@ for filename in ace_libraries:
             # Continuous-energy neutron data
             neutron = openmc.data.IncidentNeutron.from_ace(
                 table, args.metastable)
+
+            # Fission energy release data, if available
+            if args.fission_energy_release is not None:
+                fer = openmc.data.FissionEnergyRelease.from_compact_hdf5(
+                                           args.fission_energy_release, neutron)
```
Comment (paulromano) [inline scripts/openmc-ace-to-hdf5]: Hanging indent should be four spaces

**Comment by smharper** [comment]
Is it an html build?  I don't see that error.  It also looks like the numpy `Polynomial` uses the `abc.ABCMeta` metaclass so the error message doesn't make sense to me.  What do you get from `np.polynomial.Polynomial.__metaclass__`?

If the error persists, I suppose I can just remove the metaclass from Function1D.  We can just change `to_hdf5` to raise a `NotImplemented` error rather than being a full `@abstractmethod`

**Comment by paulromano** [comment]
@smharper Yes, it was from `make html`. For me, I get `np.polynomial.Polynomial.__metaclass__ = ABCMeta`, but this is because it inherits from `np.polynomial._polybase.ABCPolyBase` which is itself an abstract class with `__metaclass__ = ABCMeta`. I have to admit I'm kind of confused about all these metaclasses partially because the way they work changed from Python 2 to Python 3. I haven't been able to figure out a solution other than what you propose (reverting to raising `NotImplementedError`)

**Comment by smharper** [comment]
I wonder if it's a bug with sphinx?  Oh well.  I don't think it's worth tracking down so I've just gone for the `NotImplemented` option.

**Comment by paulromano** [comment]
Ok, I finally figured out what was going on after reading for hours about Python metaclasses. Here it goes -- when we build docs on RTD, it's not possible to have all third-party packages available because they ["can't support installing random C binaries on our system"](http://read-the-docs.readthedocs.io/en/latest/faq.html#i-get-import-errors-on-libraries-that-depend-on-c-modules). This includes things like numpy and h5py. The way we get around this is by using the nifty [mock](https://docs.python.org/3/library/unittest.mock.html) package which allows you to replace parts of your module hierarchy with mock objects. So, the real issue is that when an RTD build tries to import openmc.data.function, it hits the following line:
```Python
class Polynomial(np.polynomial.Polynomial, Function1D):
```
and sees that `type(np.polynomial.Polynomial) = MagicMock` and `type(Function1D) = type`, hence a metaclass conflict. Let's say this another way: `np.polynomial.Polynomial` is an instance of `MagicMock`. The `MagicMock` class itself is an instance of `type`, i.e., a custom metaclass. So, to fix this you can make the following change:
```diff
diff --git a/docs/source/conf.py b/docs/source/conf.py
index 4aa000f..3e226e1 100644
--- a/docs/source/conf.py
+++ b/docs/source/conf.py
@@ -28,6 +28,8 @@ MOCK_MODULES = ['numpy', 'numpy.polynomial', 'numpy.polynomial.polynomial',
                 'h5py', 'pandas', 'opencg']
 sys.modules.update((mod_name, MagicMock()) for mod_name in MOCK_MODULES)
 
+import numpy as np
+np.polynomial.Polynomial = MagicMock
 
 # If extensions (or modules to document with autodoc) are in another directory,
 # add these directories to sys.path here. If the directory is relative to the
```

After understanding metaclasses better now, I'm also confident in saying that the way we're using ABCMeta doesn't even work in Python 3. Python 3 requires that the metaclass be specified as `class(..., metaclass=...)`. It will simply ignore `__metaclass__`. This is easy to confirm; for example, I can instantiate a subclass of `Region` which doesn't implement any of the abstract methods:
```IPython
In [1]: from openmc import Region

In [2]: class A(Region):
   ...:     pass
   ...: 

In [3]: b = A()
```
The only way to have Python 2 and 3 compatible metaclasses is to use the [six](http://pythonhosted.org/six/#six.with_metaclass) library.

**Comment by smharper** [comment]
Just made the same discovery about not using metaclasses in Python3 last
night. Also, now that I understand them, metaclasses are freakin awesome.
Working on a PR now that makes good use of them.

**Comment by smharper** [comment]
I changed the conf.rst as you suggested.  Let me know if the docs build works now.

File: openmc/data/function.py [hunk 3]
```diff
@@ -9,7 +11,49 @@
                         4: 'log-linear', 5: 'log-log'}
 
 
-class Tabulated1D(object):
+class Function1D(with_metaclass(ABCMeta, object)):
```
Comment (paulromano) [inline openmc/data/function.py]: If you're ok with it, I'd suggest holding off on using six for this. If we're going to adopt a dependency on `six` we might as well use it judiciously in other areas as well (type checking for example). I think the make html issues will still be fixed even without fixing the metaclass specification here.

File: docs/source/io_formats/fission_energy.rst [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+               arrays.  The components use the 2-3 letter abbreviations
+               specified in ENDF-102 e.g. EFR for fission fragments and ENP for
+               prompt neutrons.
+
+**/<nuclide name>/**
+    Nuclides are named by concatenating their atomic symbol and mass number. For
+    example, 'U235' or 'Pu239'.  Metastable nuclides are appended with an
+    '_m' and their metastable number.  For example, 'Am242_m1'
+
+:Datasets: - **data** (*double[][][]*) -- The energy release coefficients.  The
```
Comment (paulromano) [inline docs/source/io_formats/fission_energy.rst]: Can you make the following change here:
```diff
diff --git a/docs/source/io_formats/fission_energy.rst b/docs/source/io_formats/fission_energy.rst
index 768db56..0b21a84 100644
--- a/docs/source/io_formats/fission_energy.rst
+++ b/docs/source/io_formats/fission_energy.rst
@@ -26,7 +26,8 @@ ENDF files with the
     example, 'U235' or 'Pu239'.  Metastable nuclides are appended with an
     '_m' and their metastable number.  For example, 'Am242_m1'
 
-:Datasets: - **data** (*double[][][]*) -- The energy release coefficients.  The
+:Datasets:
+           - **data** (*double[][][]*) -- The energy release coefficients.  The
              first axis indexes the component type.  The second axis specifies
              values or uncertainties.  The third axis indexes the polynomial
              order.  If the data uses the Sher-Beck format, then the last axis
```
Sphinx/rst is kind of weird. It expects a blank line before and after a bullet list, but I really only get errors when its a bullet list with only one item.

**Comment by paulromano** [comment]
Thanks for the awesomeness, @smharper!

**Pull request merged**